### DISCUSSION TITLE
Seurat v5 IntegrateLayers scVI Method Not Found

### DISCUSSION THREAD

**Discussion by mattmacd4** [satijalab/seurat#7156] [Q&A]
In working through the vignette https://satijalab.org/seurat/articles/seurat5_integration.html, getting an error when trying to run integration with scVIIntegration. 

When running this snippet:
obj <- IntegrateLayers(
  object = obj, method = scVIIntegration,
  new.reduction = "integrated.scvi",
  conda_env = "../miniconda3/envs/scvi-env", verbose = FALSE
)

I get the following error. 
Error in is_quosure(x = method) : object 'scVIIntegration' not found.

Please help!

**Comment by wudustan** [comment]
Hi, have you succesfully intalled `reticulate` and the python framework for `scvi-tools`?

From the vignette guide you posted:
>For example, scVI integration requires reticulate which can be installed from CRAN (install.packages("reticulate")) as well as scvi-tools and its dependencies installed in a conda environment. Please see scVI installation instructions [here](https://docs.scvi-tools.org/en/stable/installation.html). 

`scvi-tools` recommends you create a Python conda environment and install it that way - see the [guide](https://docs.scvi-tools.org/en/stable/installation.html) 

Then, when running the command in R, make sure that you specify a valid location for your conda environment:

```
obj <- IntegrateLayers(
  object = obj, method = scVIIntegration,
  new.reduction = "integrated.scvi",
  conda_env = "<<your conda environment location here>>", verbose = FALSE
)
```

`conda_env` argument tells the function where to look for your conda environment on your system. It looks like you copied the example but you need to change this for your own system. For example mine is: `~/miniconda3/envs/python_scvitools`

**Reply by mattmacd4** [reply]
I have followed these steps yes and substituted the path to my conda env: /opt/homebrew/anaconda3/envs/scvi-env. I repeated them and still get the same error.

I'm wondering if it has something to do with running on a Mac. I'm not sure I understand the instructions for install on Apple Silicon relating to native version of python.

**Reply by wudustan** [reply]
I'm on Intel so I can't comment specifically. Probably worth checking if you can actually get scvi-tools to run independently of R to figure out where the error arises.

**Comment by tawaya** [comment]
Same here. Also, "method = FastMNNIntegration" returned the error message "Error in is_quosure(x = method) : object 'FastMNNIntegration' not found." Other methods worked nicely (CCA, RPCA, and Harmony).

Thank you for your great work.

**Comment by Polligator** [comment]
same problem here, pretty sure my scVI is working, I have been using it for quite some time, after feeding the "cond_env" with my path of scVI conda environment, I got the error: Error in is_quosure(x = method): object 'scVIIntegration' not found, the NMF is  not working either, Error in is_quosure(x = method) : object 'FastMNNIntegration' not found, other methods like CCA, PCA worked fine.

**Reply by mmxx2022** [reply]
same. when i checked Help, it only shows 4 available methods, scVI not included.

<img width="722" alt="image" src="https://github.com/satijalab/seurat/assets/96840161/9eb6ca0a-7de8-4302-864a-b526b2682226">

**Reply by wudustan** [reply]
Can you both confirm that you have SeuratWrappers installed and loaded?

```
devtools::install_github("satijalab/seurat-wrappers", ref = "seurat5", force = TRUE, upgrade = "never")
SeuratWrappers::scVIIntegration()
SeuratWrappers::FastMNNIntegration()

```

**Reply by tawaya** [reply]
I tried reinstalling seurat-wrappers, but it returned the following message.
As to FastMNN, SeuratWrappers::RunFastMNN() exists.

`devtools::install_github("satijalab/seurat-wrappers", ref = "seurat5", force = TRUE, upgrade = "never")`

Downloading GitHub repo satijalab/seurat-wrappers@seurat5
── R CMD build ───────────────────────────────────────────────────────────────────
✔  checking for file ‘/private/var/folders/j9/lk0y89sn5cb_g3pwpbffp1gc0000gn/T/RtmpdEtWxw/remotes8166b21ebde/satijalab-seurat-wrappers-73466e3/DESCRIPTION’ ...
─  preparing ‘SeuratWrappers’:
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   Omitted ‘LazyData’ from DESCRIPTION
─  building ‘SeuratWrappers_0.3.19.tar.gz’
* installing *source* package ‘SeuratWrappers’ ...
** using staged installation
** R
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (SeuratWrappers)`

`SeuratWrappers::scVIIntegration()`
Error: 'scVIIntegration' is not an exported object from 'namespace:SeuratWrappers'

`SeuratWrappers::FastMNNIntegration()`
Error: 'FastMNNIntegration' is not an exported object from 'namespace:SeuratWrappers'

**Reply by wudustan** [reply]
So when you run `library(SeuratWrappers)` are you able to call either `scVIIntegration()` or `FastMNNIntegration()` without the double colons?

As long as the functions are available to Seurat it should work. If you can run the [example code](https://github.com/satijalab/seurat/discussions/7156#discussioncomment-5871665) below and see if it works for you. Also worth checking if restarting the session after re-installing SeuratWrappers fixes it. I still have errors that get resolved like that from time to time.

**Reply by tawaya** [reply]
Again I tried to re-install all the Seurat-related libraries and restarted the session and it worked.
The commands `SeuratWrappers::FastMNNIntegration()` and `SeuratWrappers::scVIIntegration()` were called correctly this time.

I'm on M1 Mac, so it should work for others. Thank you very much for your help!

**Comment by wudustan** [comment]
I have just tried reinstalling all prerequisites and making a new scvi-env conda environment and so far everything works.

```
devtools::install_github("mojaveazure/seurat-object", ref = "seurat5", force = TRUE, upgrade = "never")
devtools::install_github("satijalab/seurat", ref = "seurat5", force = TRUE, upgrade = "never")
devtools::install_github("satijalab/seurat-data", ref = "seurat5", force = TRUE, upgrade = "never")
devtools::install_github("satijalab/azimuth", ref = "seurat5", force = TRUE, upgrade = "never")
devtools::install_github("satijalab/seurat-wrappers", ref = "seurat5", force = TRUE, upgrade = "never")
devtools::install_github("stuart-lab/signac", ref = "seurat5", force = TRUE, upgrade = "never")

library(Seurat)
library(SeuratData)
library(SeuratWrappers)
library(Azimuth)
library(ggplot2)
library(patchwork)
options(future.globals.maxSize = 1e9)
options(Seurat.object.assay.version = "v5")

options(timeout = 1000)
InstallData("pbmcsca")
InstallData("pbmcref")

obj <- LoadData("pbmcsca")
obj <- subset(obj, nFeature_RNA > 1000)
obj <- RunAzimuth(obj, reference = "pbmcref")

obj[["RNA"]] <- split(obj[["RNA"]], f = obj$Method)

obj <- NormalizeData(obj)
obj <- FindVariableFeatures(obj)
obj <- ScaleData(obj)
obj <- RunPCA(obj)

obj <- FindNeighbors(obj, dims = 1:30, reduction = "pca")
obj <- FindClusters(obj, resolution = 2, cluster.name = "unintegrated_clusters")

obj <- RunUMAP(obj, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
DimPlot(obj, reduction = "umap.unintegrated", group.by = c("Method", "predicted.celltype.l2"))

obj <- IntegrateLayers(
  object = obj, method = FastMNNIntegration,
  new.reduction = "integrated.mnn",
  verbose = FALSE
)

obj <- IntegrateLayers(
  object = obj, method = scVIIntegration,
  new.reduction = "integrated.scvi",
  conda_env = "~/miniconda3/envs/scvi-env/", verbose = FALSE
)
```
I can confirm the reduction exists. I think a possible fix might be to double check you have installed the seuratv5 version of SeuratWrappers and load the package into your session.

My scivi-tools environment:

```
conda create -n scvi-env python=3.9
conda activate scvi-env

conda install -c conda-forge pandas numpy scanpy python-igraph leidenalg anndata scipy scvelo jax jaxlib scvi-tools
```

**Reply by MWRR** [reply]
@wudustan 
I got the error when I was trying to install Jaxlib
`conda install -c conda-forge pandas numpy scanpy python-igraph leidenalg anndata scipy scvelo jax jaxlib scvi-tools`
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - jaxlib

Current channels:

  - https://conda.anaconda.org/conda-forge/win-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.anaconda.com/pkgs/main/win-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/win-64
  - https://repo.anaconda.com/pkgs/r/noarch
  - https://repo.anaconda.com/pkgs/msys2/win-64
  - https://repo.anaconda.com/pkgs/msys2/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.

Should I install scanpy in the scvi-tools environment ?
when I tried to run 
`obj <- IntegrateLayers(
  object = obj, method = scVIIntegration,
  new.reduction = "integrated.scvi",
  conda_env = "C:/Users/h/.conda/envs/scvi-env", verbose = FALSE
)`
I got the error

> Error in py_module_import(module, convert = convert) :
ModuleNotFoundError: No module named 'scanpy'
Run `reticulate::py_last_error()` for details.



any suggestions to solve this?

**Reply by wudustan** [reply]
Hi @MWRR 

It looks like jaxlib is having issues because you're running conda in Windows? See here: 

https://jax.readthedocs.io/en/latest/developer.html

https://stackoverflow.com/questions/62585395/not-able-to-install-jaxlib


Regarding your other question, yes the conda environment where you install scvi-tools and its dependencies should also have scanpy installed. The scanpy site has some limited support if you are having issues installing it on a Windows conda environment:

https://scanpy.readthedocs.io/en/stable/installation.html
